metadata-ingestion/src/datahub/utilities/urns/_urn_base.py

```diff
@@ -0,0 +1,228 @@
+
+
+def _split_entity_id(entity_id: str) -> List[str]:
+    if not (entity_id.startswith("(") and entity_id.endswith(")")):
+        return [entity_id]
+
+    parts = []
+    start_paren_count = 1
+    part_start = 1
+    for i in range(1, len(entity_id)):
+        c = entity_id[i]
+        if c == "(":
+            start_paren_count += 1
+        elif c == ")":
+            start_paren_count -= 1
+            if start_paren_count < 0:
+                raise InvalidUrnError(f"{entity_id}, mismatched paren nesting")
+        elif c == ",":
+            if start_paren_count != 1:
+                continue
+
+            if i - part_start <= 0:
+                raise InvalidUrnError(f"{entity_id}, empty part disallowed")
+            parts.append(entity_id[part_start:i])
+            part_start = i + 1
+
+    if start_paren_count != 0:
+        raise InvalidUrnError(f"{entity_id}, mismatched paren nesting")
+
+    parts.append(entity_id[part_start:-1])
+
+    return parts
+
+
+_UrnSelf = TypeVar("_UrnSelf", bound="Urn")
+
+
+@functools.total_ordering
+class Urn:
+    """
+    URNs are globally unique identifiers used to refer to entities.
+
+    It will be in format of urn:li:<type>:<id> or urn:li:<type>:(<id1>,<id2>,...)
+
+    A note on encoding: certain characters, particularly commas and parentheses, are
+    not allowed in string portions of the URN. However, these are allowed when the urn
+    has another urn embedded within it. The main URN class ignores this possibility,
+    and assumes that the user provides a valid URN string. However, the specific URN
+    classes, such as DatasetUrn, will automatically encode these characters using
+    url-encoding when the URN is created and _allow_coercion is enabled (the default).
+    However, all from_string methods will try to preserve the string as-is, and will
+    raise an error if the string is invalid.
+    """
+
+    # retained for backwards compatibility
+    URN_PREFIX: ClassVar[str] = "urn"
+    LI_DOMAIN: ClassVar[str] = "li"
+
+    _entity_type: str
+    _entity_ids: List[str]
+
+    def __init__(self, entity_type: str, entity_id: List[str]) -> None:
+        self._entity_type = entity_type
+        self._entity_ids = entity_id
+
+        if not self._entity_ids:
+            raise InvalidUrnError("Empty entity id.")
+        for part in self._entity_ids:
+            if not part:
+                raise InvalidUrnError("Empty entity id.")
+
+    @property
+    def entity_type(self) -> str:
+        return self._entity_type
+
+    @property
+    def entity_ids(self) -> List[str]:
+        return self._entity_ids
+
+    @classmethod
+    def from_string(cls: Type[_UrnSelf], urn_str: str) -> "_UrnSelf":
+        """
+        Create a Urn from the its string representation
+
+        :param urn_str: the string representation of the Urn
+        :return: Urn of the given string representation
+
+        :raises InvalidUrnError: if the string representation is in invalid format
```

Comment:
I thought we usually use google docstring syntax. I don't care either way, but would like to be consistent